###In the project, we will analyse the data from the dataset of eBay Kleinanzeigen, a classfields section of the German eBay website to find further insight by using car listings.

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos_head=autos.head()
autos_info=autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In the output, we can see no data has null value, 5 columns are integer and 15 columns are object

In [3]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [4]:
autos.columns=['data_crawled','name','seller','offer_type','price','ab_test','vehicle_type','registration_year','gearbox','power_ps','model','obometer','registration_month','fel_type','brand','unrepaired_damage','ad_created','nr_of_pictures','postal_code','last_seen']


print(autos.columns)
print(autos.head())

Index(['data_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'obometer', 'registration_month', 'fel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')
          data_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine          

In [5]:
autos.describe(include = 'all')

data_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         obometer  registration_month fel_type       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

seller only shows 'private'. It should have either private or dealer. 
offer_type only have one type, Angebot which needs to be investigated more.
nr_of_pictures shouldn't be Nan in the top, needs to be investigated more.

In [6]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

The result shows that number of pictures has 0 in most of the rows, therefore, we decide to drop this column and sellers and offer_type

In [7]:
autos = autos.drop(["nr_of_pictures","seller","offer_type"],axis=1)
autos.describe(include='all')

data_crawled         name  price ab_test vehicle_type  \
count                 50000        50000  50000   50000        44905   
unique                48213        38754   2357       2            8   
top     2016-04-04 16:40:33  Ford_Fiesta     $0    test    limousine   
freq                      3           78   1421   25756        12859   
mean                    NaN          NaN    NaN     NaN          NaN   
std                     NaN          NaN    NaN     NaN          NaN   
min                     NaN          NaN    NaN     NaN          NaN   
25%                     NaN          NaN    NaN     NaN          NaN   
50%                     NaN          NaN    NaN     NaN          NaN   
75%                     NaN          NaN    NaN     NaN          NaN   
max                     NaN          NaN    NaN     NaN          NaN   

        registration_year  gearbox      power_ps  model   obometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fel_type       brand unrepaired_damage  \
count         50000.000000    45518       50000             40171   
unique                 NaN        7          40                 2   
top                    NaN   benzin  volkswagen              nein   
freq                   NaN    30107       10687             35232   
mean              5.723360      NaN         NaN               NaN   
std               3.711984      NaN         NaN               NaN   
min               0.000000      NaN         NaN               NaN   
25%               3.000000      NaN         NaN               NaN   
50%               6.000000      NaN         NaN               NaN   
75%               9.000000      NaN         NaN               NaN   
max              12.000000      NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [8]:
autos["price"] = (autos["price"].str.replace("$","").str.replace(",","").astype(int))

autos["price"].head()

                         

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [9]:
autos["obometer"] = (autos["obometer"].str.replace("km","").str.replace(",","").astype(int))
autos.rename({"obometer":"obometer_km"},axis=1,inplace=True)

autos["obometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: obometer_km, dtype: int64

In [10]:
autos["price"].unique().shape

(2357,)

In [11]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [12]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [13]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are 1421 price is 0 which is not normal and should be remove. Meanwhile, there's a big gap between 350000 to 999990 and the price seem unrealistic. Therefore, we decide to extract the price between 1 to 350001

In [14]:
autos["obometer_km"].unique().shape

(13,)

In [15]:
autos["obometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: obometer_km, dtype: float64

In [16]:
autos["obometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: obometer_km, dtype: int64

In [17]:
autos["obometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: obometer_km, dtype: int64

In [18]:
autos[(autos["price"].between(1,350001))]

data_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model  obometer_km  registration_month fel_type       brand  \
0      andere       150000                   3      lpg     peugeot   
1         7er       150000                   6   benzin         bmw   
2        golf        70000                   7   benzin  volkswagen   
3      fortwo        70000                   6   benzin       smart   
4       focus       150000                   7   benzin        ford   
...       ...          ...                 ...      ...         ...   
49995      q5       100000                   1   diesel        audi   
49996   astra       150000                   5   benzin        opel   
49997     500         5000                  11   benzin        fiat   
49998      a3        40000                  11   diesel        audi   
49999  vectra       150000                   1   benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[48565 rows x 17 columns]

In [19]:
autos["data_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)


2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: data_crawled, dtype: float64

The date is crawled daily in March and April 2016.

In [20]:
autos["ad_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)


2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

Ad created dates are mainly between June 2015 to April 2016

In [21]:
autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)


2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [22]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

It indicated that most car are registered in 2005 while it has some weired value such as 1000 and 9999.

As a car can't be registered before the list was seen, any year before 2016 is not accurate. It's hard to get the earliest year. Therefore, we will take the data between 1900 to 2016

In [23]:
invalid_year=(~autos["registration_year"].between(1900,2016)).sum()/autos.shape[0]
print(invalid_year)

0.03944


Luckily, it is only around 4% of the odd data.

In [24]:
print(autos[(autos["registration_year"].between(1990,2016))])

              data_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price  ab_test vehic

In [25]:
print(autos["registration_year"].value_counts(normalize=True))

2000    0.06708
2005    0.06030
1999    0.06000
2004    0.05474
2003    0.05454
         ...   
1931    0.00002
1929    0.00002
1001    0.00002
9996    0.00002
1952    0.00002
Name: registration_year, Length: 97, dtype: float64


In [26]:
autos["brand"].unique()
autos["brand"].value_counts().head(20)
over_5 = autos["brand"].value_counts()/autos.shape[0]
print(over_5)

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

The top 5 brand is German company and it's over 50% of the whole list. Many brands are lower than 5% and we will focus on the brand which has high proportion.

In [27]:
brand_counts = autos["brand"].value_counts(normalize = True)
common_brands = brand_counts[brand_counts>0.05].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [28]:
brand_price = {}
for row in common_brands:
    brand_only = autos[autos["brand"] == row]
    mean_price = brand_only["price"].mean()
    brand_price[row] = int(mean_price)
    
print(brand_price)

{'volkswagen': 6384, 'opel': 5106, 'bmw': 8252, 'mercedes_benz': 29511, 'audi': 8965, 'ford': 7105}


In the result, it shows that Mercedes is the most expensive and Opel is the most affordable brand.

In [30]:
milage_counts = autos["obometer_km"].value_counts(normalize = True)
print(milage_counts)

150000    0.64848
125000    0.10340
100000    0.04338
90000     0.03514
80000     0.02872
70000     0.02460
60000     0.02328
50000     0.02054
5000      0.01934
40000     0.01638
30000     0.01578
20000     0.01568
10000     0.00528
Name: obometer_km, dtype: float64


In [32]:
brand_milage = {}
for row in common_brands:
    brand_only = autos[autos["brand"]== row]
    mean_milage = brand_only["obometer_km"].mean()
    brand_milage[row]= float(mean_milage)
    
print(brand_milage)

{'volkswagen': 128955.27276129878, 'opel': 129298.66324848929, 'bmw': 132521.64302818198, 'mercedes_benz': 130886.14279678918, 'audi': 129643.9411627364, 'ford': 124131.93446392642}


In [35]:
brand_price_series = pd.Series(brand_price).sort_values(ascending = False)
brand_milage_series = pd.Series(brand_milage).sort_values(ascending=False)
brand_price_dataframe = pd.DataFrame(brand_price_series, columns = ['mean_price'])
brand_milage_dataframe = pd.DataFrame(brand_milage_series, columns = ['mean_milage'])
print(brand_price_dataframe)
print(brand_milage_dataframe)

               mean_price
mercedes_benz       29511
audi                 8965
bmw                  8252
ford                 7105
volkswagen           6384
opel                 5106
                 mean_milage
bmw            132521.643028
mercedes_benz  130886.142797
audi           129643.941163
opel           129298.663248
volkswagen     128955.272761
ford           124131.934464


In [41]:
autos["last_seen"]=autos["last_seen"].str[:10].str.replace("-","").astype(int)
print(autos["last_seen"])

0        20160406
1        20160406
2        20160406
3        20160315
4        20160401
           ...   
49995    20160401
49996    20160402
49997    20160404
49998    20160405
49999    20160406
Name: last_seen, Length: 50000, dtype: int64
